File: polyglot/go/layer_size_calculator.go
package main

import (
	"bufio"
	"bytes"
	"flag"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"strconv"
	"strings"
)

// LayerInfo tracks size after each instruction
type LayerInfo struct {
	Command     string
	Cumulative  int64 // bytes
	Estimated   int64 // estimated layer size
}

// DockerfileState holds parsing context
type DockerfileState struct {
	Layers      []LayerInfo
	BaseImage    string
	CurrentSize  int64
	WorkingDir   string
	Args         map[string]string
	Env          map[string]string
	MultiLineCmd string
}

// Configurable thresholds for warnings
var (
	WarningThreshold = 100 * 1024 * 1024 // 100MB
	CriticalThreshold = 500 * 1024 * 1024 // 500MB
)

// ParseDockerfile reads and parses a Dockerfile path
func ParseDockerfile(path string) (*DockerfileState, error) {
	f, err := os.Open(path)
	if err != nil {
		return nil, fmt.Errorf("open file: %w", err)
	}
	defer f.Close()

	state := &DockerfileState{
		Layers:    make([]LayerInfo, 0),
		Args:      make(map[string]string),
		Env:       make(map[string]string),
		MultiLineCmd: "",
	}

	scanner := bufio.NewScanner(f)
	lineNum := 1

	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		
		// Skip empty lines and comments
		if line == "" || strings.HasPrefix(line, "#") {
			continue
		}

		// Handle multi-line commands (RUN | COPY --from=...)
		if strings.Contains(line, "|") && !strings.Contains(line, "FROM") {
			state.MultiLineCmd += line + "\n"
			continue
		} else if state.MultiLineCmd != "" {
			line = state.MultiLineCmd + line
			state.MultiLineCmd = ""
		}

		parts := parseInstruction(line)
		if parts == nil {
			continue
		}

		cmd, args, ok := parts[0], parts[1], true
		if !ok {
			cmd = line
			args = ""
		}

		// Update state for ARG/ENV
		if cmd == "ARG" {
			parts := strings.SplitN(args, "=", 2)
			if len(parts) == 2 {
				state.Args[parts[0]] = parts[1]
			}
			continue
		}

		if cmd == "ENV" {
			parts := strings.SplitN(args, "=", 2)
			if len(parts) == 2 {
				state.Env[parts[0]] = parts[1]
			}
			continue
		}

		// Handle FROM instruction
		if cmd == "FROM" {
			state.BaseImage = strings.TrimSpace(args)
			state.CurrentSize = 0 // Reset for new image
			state.Layers = []LayerInfo{}
			continue
		}

		// Handle WORKDIR
		if cmd == "WORKDIR" {
			state.WorkingDir = strings.TrimPrefix(strings.TrimSpace(args), "/")
			continue
		}

		// Calculate layer size for other instructions
		size := estimateLayerSize(cmd, args)
		state.CurrentSize += size
		
		layerInfo := LayerInfo{
			Command:     cmd,
			Cumulative:  state.CurrentSize,
			Estimated:   size,
		}

		// Add to layers slice (insert at beginning for chronological order)
		state.Layers = append([]LayerInfo{layerInfo}, state.Layers...)

		// Check thresholds
		if state.CurrentSize >= CriticalThreshold {
			fmt.Printf("CRITICAL: Layer %s now at %.2f MB\n", cmd, float64(state.CurrentSize)/1024/1024)
		} else if state.CurrentSize >= WarningThreshold {
			fmt.Printf("WARNING: Layer %s now at %.2f MB\n", cmd, float64(state.CurrentSize)/1024/1024)
		}

		// Handle multi-line continuation
		if strings.HasSuffix(args, "|") || strings.Contains(cmd, "RUN ") && strings.HasSuffix(strings.TrimSpace(args), "|") {
			state.MultiLineCmd = line + "\n"
			continue
		}
	}

	return state, scanner.Err()
}

// parseInstruction splits a Dockerfile instruction into command and arguments
func parseInstruction(line string) []string {
	line = strings.TrimPrefix(line, "#") // Remove inline comments
	parts := strings.SplitN(line, " ", 2)
	
	if len(parts) < 1 {
		return nil
	}

	cmd := parts[0]
	args := ""
	if len(parts) == 2 {
		args = parts[1]
	}

	// Normalize command name (remove trailing whitespace)
	cmd = strings.TrimSpace(cmd)

	return []string{cmd, args}
}

// estimateLayerSize calculates estimated size contribution of an instruction
func estimateLayerSize(cmd string, args string) int64 {
	switch cmd {
	case "FROM":
		if len(args) > 0 {
			// Estimate base image size (rough approximation)
			image := strings.ToLower(strings.TrimSpace(args))
			return estimateBaseImageSize(image)
		}
		return 1024 * 1024 // Default 1MB for unknown

	case "RUN":
		if len(args) == 0 {
			return 512 * 1024 // Default 512KB for empty RUN
		}
		
		// Parse command and estimate based on common patterns
		cmdParts := strings.Fields(args)
		if len(cmdParts) == 0 {
			return 512 * 1024
		}

		firstWord := cmdParts[0]
		
		// apt-get install/update
		if firstWord == "apt-get" && (strings.Contains(args, "install") || strings.Contains(args, "update")) {
			return estimateAPTSize(args)
		}

		// apk add
		if firstWord == "apk" && strings.Contains(args, "add") {
			return 512 * 1024 // Conservative estimate for Alpine packages
		}

		// yum/dnf install
		if (firstWord == "yum" || firstWord == "dnf") && strings.Contains(args, "install") {
			return 768 * 1024
		}

		// pip/pip3 install
		if firstWord == "pip" || firstWord == "pip3" {
			pkgCount := countPipPackages(args)
			return int64(pkgCount) * 50 * 1024 // ~50KB per package average
		}

		// npm install (node.js)
		if firstWord == "npm" && strings.Contains(args, "install") {
			deps := countNPMDependencies(args)
			return int64(deps) * 128 * 1024 // ~125KB per dependency average
		}

		// yarn install (node.js)
		if firstWord == "yarn" && strings.Contains(args, "install") {
			deps := countYARNDependencies(args)
			return int64(deps) * 96 * 1024 // ~95KB per dependency average
		}

		// curl/wget download (common in multi-stage builds)
		if firstWord == "curl" || firstWord == "wget" {
			return estimateDownloadSize(args)
		}

		// Copy from cache layers (--from=cache:tag)
		if strings.Contains(args, "--from=") && !strings.Contains(cmd, "COPY --") {
			return 256 * 1024 // Conservative for COPY with FROM
		}

		// Default estimate for RUN commands
		return 768 * 1024

	case "COPY":
		if len(args) == 0 {
			return 512 * 1024
		}

		// Check if copying from cache layer
		if strings.Contains(args, "--from=") && !strings.Contains(cmd, "COPY --") {
			cacheLayers := countCacheLayers(args)
			return int64(cacheLayers) * 512 * 1024 // ~500KB per cached layer
		}

		// Check if copying from current image (--from=...)
		if strings.Contains(args, "--from=") && strings.Contains(cmd, "COPY --") {
			return 768 * 1024
		}

		// Estimate file sizes based on patterns
		files := extractFilePaths(args)
		
		// Check for common large files
		largeFiles := countLargeFiles(files)
		if largeFiles > 0 {
			return int64(largeFiles) * 256 * 1024 // ~250KB per large file estimate
		}

		// Default estimate for regular COPY
		return 256 * 1024

	case "ADD":
		// ADD is similar to COPY but with additional features
		if len(args) == 0 {
			return 768 * 1024
		}

		files := extractFilePaths(args)
		
		// Check for tar archives (unpacked during build)
		tarFiles := countTarFiles(files)
		if tarFiles > 0 {
			return int64(tarFiles) * 512 * 1024 // ~500KB per archive estimate
		}

		// Check for URL downloads (--from-url=...)
		urls := countURLs(args)
		if urls > 0 {
			return int64(urls) * 1024 * 1024 // ~1MB per URL (conservative)
		}

		// Default estimate for ADD
		return 512 * 1024

	case "ENV", "ARG":
		// ENV and ARG have minimal size impact
		return 64 * 1024 // ~64KB for environment/argument metadata

	case "WORKDIR", "LABEL", "EXPOSE", "CMD", "ENTRYPOINT":
		// Minimal overhead instructions
		return 32 * 1024 // ~32KB

	default:
		// Unknown instruction - conservative estimate
		return 512 * 1024
	}
}

// estimateBaseImageSize returns estimated size for common base images
func estimateBaseImageSize(image string) int64 {
	image = strings.ToLower(strings.TrimSpace(image))

	switch image {
	case "alpine", "alpine:3.18", "alpine:latest":
		return 50 * 1024 * 1024 // ~50MB for Alpine
	case "debian", "debian:bullseye", "debian:buster":
		return 128 * 1024 * 1024 // ~128MB for Debian
	case "ubuntu", "ubuntu:focal", "ubuntu:jammy":
		return 768 * 1024 * 1024 // ~750MB for Ubuntu
	case "centos", "centos:stream9", "centos:latest":
		return 384 * 1024 * 1024 // ~380MB for CentOS
	case "rhel", "rockylinux", "almalinux":
		return 512 * 1024 * 1024 // ~500MB for RHEL family
	case "fedora", "fedora:latest":
		return 384 * 1024 * 1024 // ~380MB for Fedora
	case "suse", "opensuse", "opensuse-leap":
		return 256 * 1024 * 1024 // ~250MB for SUSE
	case "busybox", "busybox:latest":
		return 3 * 1024 * 1024 // ~3MB for BusyBox
	case "scratch":
		return 64 * 1024 // ~64KB for scratch (empty)
	default:
		// Default estimate for unknown images
		return 256 * 1024 * 1024 // ~250MB conservative default
	}
}

// estimateAPTSize estimates size for apt-get operations
func estimateAPTSize(args string) int64 {
	// Count packages being installed/updated
	pkgCount := countAPTPackages(args)
	
	if pkgCount > 0 {
		return int64(pkgCount) * 256 * 1024 // ~250KB per package average
	}

	// Default for apt-get update (no packages specified)
	return 512 * 1024
}

// countAPTPackages counts packages in apt-get commands
func countAPTPackages(args string) int {
	var count int
	
	// Match "package" or "packages" followed by names
	re := regexp.MustCompile(`(apt|apt-get)\s+(install|update|upgrade)\s+([^&\n]+)` + 
			regexp.QuoteMeta("&&") + `[^&\n]+`)
	
	matches := re.FindAllString(args, -1)
	for _, match := range matches {
		parts := strings.Fields(match)
		if len(parts) >= 3 {
			cmd := parts[0] + " " + parts[1]
			if cmd == "apt-get install" || cmd == "apt install" {
				count += countPackageNames(parts[2])
			} else if cmd == "apt-get update" || cmd == "apt update" {
				count++ // Update adds some overhead
			}
		}
	}

	return count
}

// countPackageNames counts individual package names
func countPackageNames(names string) int {
	parts := strings.Fields(names)
	var count int
	
	for _, part := range parts {
		if !strings.Contains(part, "&&") && 
		   !strings.Contains(part, ";") &&
		   !strings.Contains(part, "=") {
			count++
		}
	}

	return count
}

// countPipPackages counts pip packages in install commands
func countPipPackages(args string) int {
	var count int
	
	parts := strings.Fields(args)
	for i, part := range parts {
		if (part == "pip" || part == "pip3") && 
		   (parts[i+1] == "install" || parts[i+1] == "freeze") {
			count += countPackageNames(parts[i+2])
		}
	}

	return count
}

// countNPMDependencies counts npm dependencies in install commands
func countNPMDependencies(args string) int {
	var count int
	
	parts := strings.Fields(args)
	for i, part := range parts {
		if (part == "npm" || part == "yarn") && 
		   (parts[i+1] == "install" || parts[i+1] == "add") {
			count += countPackageNames(parts[i+2])
		}
	}

	return count
}

// countYARNDependencies counts yarn dependencies in install commands
func countYARNDependencies(args string) int {
	var count int
	
	parts := strings.Fields(args)
	for i, part := range parts {
		if (part == "yarn") && 
		   (parts[i+1] == "install" || parts[i+1] == "add") {
			count += countPackageNames(parts[i+2])
		}
	}

	return count
}

// estimateDownloadSize estimates size for curl/wget downloads
func estimateDownloadSize(args string) int64 {
	var total int64
	
	parts := strings.Fields(args)
	for i, part := range parts {
		if (part == "curl" || part == "wget") && 
		   (parts[i+1] == "-o" || parts[i+1] == "--output-dir=" || parts[i+1] == "-O") {
			total++ // Conservative estimate per download
		}
	}

	return total * 512 * 1024 // ~500KB per download (conservative)
}

// countCacheLayers counts --from= cache layers in COPY/ADD instructions
func countCacheLayers(args string) int {
	count := strings.Count(args, "--from=")
	return count
}

// extractFilePaths extracts file paths from COPY/ADD arguments
func extractFilePaths(args string) []string {
	var files []string
	
	// Remove flags and options
	cleanArgs := args
	for re := range map[string]func(string) string{
		`--from=`: func(s string) string {
			idx := strings.Index(s, "--from

File: polyglot/rust/layer_size_calculator.rs
use std::collections::HashMap;
use std::env;
use std::fs::{self, File};
use std::io::{BufRead, BufReader, Write};
use std::path::Path;
use std::process;

/// Configuration for layer size estimation
struct Config {
    /// Default overhead per RUN instruction (apt-get/dnf cache + metadata)
    run_overhead: u64,
    /// Overhead per COPY/ADD (metadata, directory creation)
    copy_overhead: u64,
    /// Minimum file size to report for COPY/ADD
    min_file_size: u64,
    /// Warning threshold in bytes
    warning_threshold: u64,
    /// Critical threshold in bytes
    critical_threshold: u64,
}

impl Default for Config {
    fn default() -> Self {
        Self {
            run_overhead: 50_000_000,      // ~50MB per RUN
            copy_overhead: 1024,           // ~1KB metadata
            min_file_size: 1024 * 1024,    // 1MB minimum to report
            warning_threshold: 100_000_000, // 100MB warning
            critical_threshold: 500_000_000, // 500MB critical
        }
    }
}

/// Represents a parsed Dockerfile instruction with its estimated size impact
struct Layer {
    line_number: usize,
    instruction: String,
    raw_line: String,
    estimated_size_bytes: u64,
    file_count: u32,
    notes: Vec<String>,
}

/// Result of parsing a Dockerfile
pub struct ParseResult {
    layers: Vec<Layer>,
    total_size_estimate: u64,
    warnings: Vec<(usize, String)>,
    criticals: Vec<(usize, String)>,
}

impl ParseResult {
    pub fn new() -> Self {
        Self {
            layers: Vec::new(),
            total_size_estimate: 0,
            warnings: Vec::new(),
            criticals: Vec::new(),
        }
    }

    pub fn with_layers(mut self, layers: Vec<Layer>) -> Self {
        let mut total = 0;
        for layer in &layers {
            total += layer.estimated_size_bytes;
        }
        self.total_size_estimate = total;
        self.layers = layers;
        self
    }

    pub fn with_warning(mut self, line: usize, msg: String) -> Self {
        self.warnings.push((line, msg));
        self
    }

    pub fn with_critical(mut self, line: usize, msg: String) -> Self {
        self.criticals.push((line, msg));
        self
    }
}

/// Estimate size contribution of a single instruction
fn estimate_instruction_size(
    instruction: &str,
    args: &str,
    config: &Config,
) -> (u64, u32, Vec<String>) {
    let mut notes = Vec::new();
    
    // FROM - starts fresh layer
    if instruction.to_uppercase() == "FROM" {
        return (10_000, 0, vec!["Base image layer".to_string()]);
    }

    // COPY/ADD - extract file sizes from args
    let mut file_size = config.copy_overhead;
    let mut files = 0u32;
    
    if instruction.to_uppercase() == "COPY" || instruction.to_uppercase() == "ADD" {
        // Parse source paths and estimate sizes
        let parts: Vec<&str> = args.split_whitespace().collect();
        
        for part in &parts[1..] {
            // Check if this looks like a file path (not a flag)
            if !part.starts_with('-') && !part.contains(':') || 
               part.contains('/') {
                files += 1;
                
                // Estimate file size based on common patterns
                let estimated = if part.contains("tar") || part.contains("gz") {
                    // Archives are compressed, estimate decompressed size
                    match part.split('.').last() {
                        Some("tar.gz") | Some("tgz") => *part.len() as u64 * 3,
                        Some("tar") => *part.len() as u64 * 5,
                        _ => *part.len() as u64 * 2,
                    }
                } else if part.contains("deb") || part.contains("rpm") {
                    // Package files - estimate based on common sizes
                    match part.split('.').last() {
                        Some("deb") | Some("rpm") => 5_000_000,
                        _ => *part.len() as u64 * 2,
                    }
                } else if part.contains("curl") || part.contains("wget") {
                    // Downloaded content - estimate from URL length hint
                    match part.split('.').last() {
                        Some(ext) if ext == "gz" | ext == "bz2" => *part.len() as u64 * 10,
                        _ => *part.len() as u64 * 5,
                    }
                } else {
                    // Regular file - estimate from path length hint
                    part.len() as u64 * 3
                };
                
                file_size += estimated;
            }
        }
        
        return (file_size.max(10_000), files, vec!["COPY/ADD layer".to_string()]);
    }

    // RUN - command execution overhead
    if instruction.to_uppercase() == "RUN" {
        let estimated = config.run_overhead;
        
        // Check for common patterns that add significant size
        let mut cmd_notes = vec!["RUN layer".to_string()];
        
        if args.contains("apt-get install") || args.contains("dnf install") {
            cmd_notes.push("Package manager install - adds cache + packages".to_string());
        } else if args.contains("apk add") {
            cmd_notes.push("Alpine package manager".to_string());
        } else if args.contains("yum install") {
            cmd_notes.push("RHEL/CentOS package manager".to_string());
        } else if args.contains("pip install") || args.contains("npm install") || 
                  args.contains("cargo add") {
            cmd_notes.push("Language dependency installation - may add significant size".to_string());
        } else if args.contains("curl") || args.contains("wget") {
            cmd_notes.push("Network download - consider caching or multi-stage build".to_string());
        } else if args.contains(".tar.gz") || args.contains(".tgz") {
            cmd_notes.push("Archives included - consider extracting in RUN then copying only needed files".to_string());
        }

        return (estimated, 0, cmd_notes);
    }

    // ENV/ARG/LABEL - minimal overhead
    if instruction.to_uppercase() == "ENV" || 
       instruction.to_uppercase() == "ARG" ||
       instruction.to_uppercase() == "LABEL" {
        let estimated = if args.len() > 50 { 1_000 } else { 256 };
        return (estimated, 0, vec!["Metadata layer".to_string()]);
    }

    // WORKDIR/USER/EXPOSE - negligible
    if instruction.to_uppercase() == "WORKDIR" || 
       instruction.to_uppercase() == "USER" ||
       instruction.to_uppercase() == "EXPOSE" {
        return (256, 0, vec!["Metadata layer".to_string()]);
    }

    // Default - estimate from line length
    let estimated = args.len().max(1) as u64 * 3;
    return (estimated, 0, vec!["Unknown/Other layer".to_string()]);
}

/// Parse a Dockerfile and calculate layer sizes
pub fn parse_dockerfile(path: &Path, config: Option<&Config>) -> ParseResult {
    let cfg = config.unwrap_or(&Config::default());
    
    let mut result = ParseResult::new();
    let mut total_size = 0u64;
    
    // Read file content first to handle multi-line commands
    let content = fs::read_to_string(path).unwrap_or_default();
    let lines: Vec<&str> = content.lines().collect();
    
    for (line_num, line) in lines.iter().enumerate() {
        let line_num = line_num + 1;
        
        // Skip empty lines and comments
        if line.trim().is_empty() || line.trim().starts_with('#') {
            continue;
        }

        // Handle multi-line commands (like pip install, curl | tar)
        let mut current_line = String::new();
        let mut is_continuation = false;
        
        for ch in line.chars() {
            if ch == '\\' && !is_continuation {
                // Line continuation - peek at next line
                if line_num < lines.len() - 1 {
                    current_line.push(ch);
                    is_continuation = true;
                    
                    let next_line = lines[line_num + 1].trim();
                    if !next_line.is_empty() && !next_line.starts_with('#') {
                        current_line.push_str(next_line.trim());
                    } else {
                        // End of continuation
                        break;
                    }
                } else {
                    current_line.push(ch);
                }
            } else {
                current_line.push(ch);
            }
        }
        
        let trimmed = current_line.trim();
        if !trimmed.is_empty() && !trimmed.starts_with('#') {
            // Parse instruction and arguments
            let parts: Vec<&str> = trimmed.split_whitespace().collect();
            
            if parts.is_empty() || parts[0].is_empty() {
                continue;
            }

            let instruction = parts[0];
            let args = if parts.len() > 1 { &parts[1..].join(" ") } else { "" };
            
            // Estimate size for this layer
            let (size, files, notes) = estimate_instruction_size(instruction, args, cfg);
            
            // Check thresholds
            if size >= cfg.critical_threshold {
                result.with_critical(line_num, format!(
                    "Large layer detected: {} bytes",
                    human_readable(size)
                ));
            } else if size >= cfg.warning_threshold {
                result.with_warning(line_num, format!(
                    "Consider optimizing this layer: {}",
                    instruction
                ));
            }

            // Add to total
            total_size += size;
            
            let mut layer = Layer {
                line_number: line_num,
                instruction: instruction.to_string(),
                raw_line: trimmed.to_string(),
                estimated_size_bytes: size,
                file_count: files,
                notes,
            };

            // Add notes to result if any warnings/criticals in this layer
            for note in &notes {
                if !note.is_empty() && 
                   (size >= cfg.warning_threshold || note.contains("consider") || 
                    note.contains("optimize")) {
                    let warning_msg = format!("Layer {}: {}", line_num, note);
                    result.with_warning(line_num, warning_msg);
                }
            }

            result.layers.push(layer);
        }
        
        current_line.clear();
        is_continuation = false;
    }

    // Sort layers by line number (should already be sorted, but ensure it)
    result.layers.sort_by_key(|l| l.line_number);

    result.with_layers(result.layers)
}

/// Format bytes into human-readable format
fn human_readable(bytes: u64) -> String {
    const KB: u64 = 1024;
    const MB: u64 = 1024 * 1024;
    const GB: u64 = 1024 * 1024 * 1024;

    if bytes >= GB {
        format!("{:.2} GB", bytes as f64 / GB as f64)
    } else if bytes >= MB {
        format!("{:.2} MB", bytes as f64 / MB as f64)
    } else if bytes >= KB {
        format!("{:.1} KB", bytes as f64 / KB as f64)
    } else {
        format!("{} B", bytes)
    }
}

/// Summary report for a ParseResult
pub fn summary_report(result: &ParseResult, config: &Config) -> String {
    let mut output = String::new();
    
    output.push_str(&format!(
        "=== Layer Size Analysis\n{}\n",
        "=".repeat(40)
    ));

    // Overall statistics
    output.push_str(&format!("Total estimated size: {}\n", human_readable(result.total_size_estimate)));
    output.push_str(&format!("Number of layers: {}\n\n", result.layers.len()));

    if !result.layers.is_empty() {
        // Layer-by-layer breakdown (top 20 by size)
        let mut sorted_layers = &mut result.layers.clone();
        sorted_layers.sort_by(|a, b| b.estimated_size_bytes.cmp(&a.estimated_size_bytes));
        
        output.push_str("Layer Breakdown:\n");
        output.push_str("-".repeat(40).as_str());

        let display_count = std::cmp::min(sorted_layers.len(), 20);
        for (i, layer) in sorted_layers.iter().take(display_count).enumerate() {
            let rank = i + 1;
            let pct = if result.total_size_estimate > 0 {
                format!("{:.1}%", 
                    (layer.estimated_size_bytes as f64 / result.total_size_estimate as f64) * 100.0)
            } else {
                "N/A".to_string()
            };

            output.push_str(&format!(
                "\n{}: {} ({})\n",
                rank,
                layer.instruction,
                human_readable(layer.estimated_size_bytes)
            ));
            
            if !layer.notes.is_empty() {
                for note in &layer.notes {
                    output.push_str(&format!("  • {}\n", note));
                }
            }

            // Highlight large layers
            if layer.estimated_size_bytes >= config.warning_threshold {
                output.push_str("    ⚠️  Large layer - consider optimization\n");
            } else if layer.estimated_size_bytes >= config.critical_threshold {
                output.push_str("    🔴 Critical size!\n");
            }
        }

        if sorted_layers.len() > 20 {
            output.push_str(&format!("\n... and {} more layers\n", 
                sorted_layers.len() - 20));
        }
    }

    // Warnings summary
    if !result.warnings.is_empty() {
        output.push_str(&format!(
            "\n=== Warnings ({}) ===\n",
            result.warnings.len()
        ));
        for (line, msg) in &result.warnings {
            output.push_str(&format!("Line {}: {}\n", line, msg));
        }
    }

    // Criticals summary
    if !result.criticals.is_empty() {
        output.push_str(&format!(
            "\n=== Critical Issues ({}) ===\n",
            result.criticals.len()
        ));
        for (line, msg) in &result.criticals {
            output.push_str(&format!("Line {}: {}\n", line, msg));
        }
    }

    // Recommendations
    if !result.warnings.is_empty() || !result.criticals.is_empty() {
        output.push_str("\n=== Recommendations ===\n");
        
        let mut recommendations = Vec::new();
        
        if result.layers.iter().any(|l| l.estimated_size_bytes >= config.warning_threshold) {
            recommendations.push("Consider using multi-stage builds to reduce final image size".to_string());
        }
        
        if result.layers.iter().any(|l| l.instruction == "RUN") {
            recommendations.push("Combine RUN commands when possible to reduce layer count".to_string());
        }
        
        if result.layers.iter().any(|l| l.file_count > 0) {
            recommendations.push("Use COPY --from=previous_stage for dependencies instead of copying from host".to_string());
        }

        for rec in &recommendations {
            output.push_str(&format!("• {}\n", rec));
        }
    }

    output
}

/// Write summary to a file
pub fn write_report(result: &ParseResult, config: &Config, path: &Path) -> std::io::Result<()> {
    let report = summary_report(result, config);
    
    // Create parent directories if needed
    if let Some(parent) = path.parent() {
        if !parent.as_os_str().is_empty() {
            fs::create_dir_all(parent)?;
        }
    }
    
    File::create(path)?.write_all(report.as_bytes())?;
    Ok(())
}

/// Main entry point for the tool
pub fn main() -> std::io::Result<()> {